// libqa/exports/HTML.js
/*
 * QA HTML Export class
 * Author: Stéphane Desneux <sdx@kooltux.org>
 * Date: 2013-02-11
 *
 * Copyright 2013, Intel Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 */

var _=require('underscore');
var util=require('util');
var log4js=require("log4js");
var path=require("path");
var fs=require("fs");
var jade=require("jade");

var logger=log4js.getLogger("qaexphtml");

var QA=require('libqa');

var DEBUG=false;

/* ------------------ plugin default options ---------------- */
var DEFAULTS={
};

/* ------------------ plugin class ---------------- */

var Plugin=module.exports=function(opts) {
	_.extend(this,
		_.extend(DEFAULTS,opts || {})
	);
}

Plugin.prototype.setId=function(id) {
	DEBUG && logger.debug("Plugin id: "+id);
	this.id=id;
}

Plugin.prototype.getFilename=function() {
	return "export_"+this.id+".html";
}

Plugin.prototype.execute=function(testsuite,cb) {
	var template=path.join(__dirname,"HTML","template.jade");

	// compile template
	var jadefn = jade.compile(fs.readFileSync(template),{
		filename: path.resolve(template)
	});

	// build data tree
	var suite={
		name: testsuite.getName(),
		sets: []
	};

	testsuite.forEachTestset(function(tset) {
		var set={
			component: tset.getComponentName(),
			name: tset.getId(),
			cases: []
		};
		
		tset.forEachTestcase(function(tcase) {
			var obj=tcase.toExportable();
			set.cases.push(obj);
		},this);

		suite.sets.push(set);
	},this);

	var template_params={
		suite: suite
	};

	// execute template
//DEBUG && logger.debug("Template params: "+util.inspect(template_params,false,null));
	cb(jadefn(template_params));
}
