// libqa/index.js
/*
 * QA library namespace module
 * Author: Stéane Desneux <sdx@kooltux.org>
 * Date: 2013-01-10
 *
 * Copyright 2013, Intel Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 */

exports.GLOBALS={
	hwtarget_rundir: "/qa",
	hwtarget_packdir: "/usr/share/tests/common",
	hwtarget_tempdir: "/tmp",
	tarball_topdir: "TESTDIR",
	local_tempdir: "/tmp/libqa"
};

exports.utils=require("./utils.js");

exports.Config=require("./Config.js");

exports.Run=require("./Run.js");
exports.Export=require("./Export.js");
exports.Import=require("./Import.js");
exports.Executor=require("./Executor.js");

exports.Database=require("./Database.js");
exports.Testsuite=require("./Testsuite.js");
exports.Testset=require("./Testset.js");
exports.Testcase=require("./Testcase.js");

exports.Services=require("./Services.js");

exports.TestKit=require("./TestKit.js");
